hack/log-dump.sh

```diff
@@ -0,0 +1,160 @@
+  exit 1
+elif [[ -z "${AZURE_SSH_PRIVATE_KEY_FILE:-}" ]]; then
+  echo "AZURE_SSH_PRIVATE_KEY_FILE not set, but required to SSH into nodes"
+  exit 1
+elif [[ -z "${AZURE_SSH_USER:-}" ]]; then
+  echo "AZURE_SSH_USER not set, but required to SSH into nodes"
+  exit 1
+elif [[ -z "${ARTIFACTS:-}" ]]; then
+  echo "ARTIFACTS not set, defaulting to /workspace/_artifacts"
+  export ARTIFACTS="/workspace/_artifacts"
+fi
+
+# Prefixes of log files we are collecting
+readonly master_logfiles=(
+  "kube-apiserver"
+  "kube-scheduler"
+  "kube-controller-manager"
+  "kube-addon-manager"
+  "cloud-controller-manager"
+)
+readonly node_logfiles=(
+  "csi-azuredisk"
+  "csi-azurefile"
+  "kube-proxy"
+)
+readonly systemd_services=(
+  "kubelet"
+  "docker"
+  "etcd"
+)
+
+# SSH into host $1 and execute command $2 inside the host
+function log-dump-ssh() {
+  local -r host="${1}"
+  local -r cmd="${2}"
+  if [[ -z "${host}" ]] || [[ -z "${cmd}" ]]; then
+    return
+  fi
+  ssh -oLogLevel=quiet -oConnectTimeout=30 -oStrictHostKeyChecking=no -i "${AZURE_SSH_PRIVATE_KEY_FILE}" "${AZURE_SSH_USER}@${host}" "${cmd}" || true
+}
+
+# Copy each file in $1 from host $2 to destination folder $3 via scp
+function copy-log-from-node() {
+  local -r scp_files="${1}"
+  local -r host="${2}"
+  local -r destination="${3}"
+  scp -T -oLogLevel=quiet -oConnectTimeout=30 -oStrictHostKeyChecking=no -i "${AZURE_SSH_PRIVATE_KEY_FILE}" "${AZURE_SSH_USER}@${host}:${scp_files}" "${destination}" || true
+}
+
+# Return the URL / IP address of node $1.
+# $2 is true if $1 is master.
+function get-host() {
+  local -r node_name="${1}"
+  local -r is_master="${2:-"false"}"
+  if [[ "${is_master}" == "true" ]]; then
+    # Remove "https://" from the URL
+    echo "$(kubectl config view --minify -ojsonpath="{.clusters[0].cluster.server}" | sed -e "s/https\:\/\///")"
+  else
+    echo "$(kubectl get nodes -ojsonpath={.items[?\(@.metadata.name==\"${node_name}\"\)].status.addresses[?\(@.type==\"ExternalIP\"\)].address})"
+  fi
+}
+
+# Dump log files from each master to a local folder
+function dump-master-log() {
+  local -r master_names=( $(kubectl get nodes -l kubernetes.io/role=master -ojsonpath={.items[*].metadata.name}) )
+  for master_name in "${master_names[@]}"; do
+    local master_host="$(get-host "${master_name}" "true")"
+    if [[ -z "${master_host}" ]]; then
+      echo "${master_name} does not have an external IP for the script to SSH into. Skipping..."
+      return
+    fi
+
+    dump-log "true" "${master_host}" "${master_name}"
+  done
+}
+
+# Dump log files from each node to a local folder
+function dump-node-log() {
+  local -r node_names=( $(kubectl get nodes -l kubernetes.io/role!=master -ojsonpath={.items[*].metadata.name}) )
+  for node_name in "${node_names[@]}"; do
+    local node_host="$(get-host "${node_name}" "false")"
+    if [[ -z "${node_host}" ]]; then
+      echo "${node_name} does not have an external IP for the script to SSH into. Skipping..."
```

Comment:
same here for nodes. And there're conflicts for SLB+outbound rules+VMSS as public IPs are not allowed on VMSS. Could we get them from master proxy, or by running new Pods on each node?